File: packages/less/src/less/tree/javascript.test.js
import { describe, it, expect } from 'vitest';
import JavaScript from './javascript';
import Dimension from './dimension';
import Quoted from './quoted';
import Anonymous from './anonymous';

describe('JavaScript', () => {
    const mockContext = {
        javascriptEnabled: true,
        frames: [
            {
                variables: () => ({}),
                variable: () => null
            }
        ]
    };

    describe('constructor', () => {
        it('should initialize with provided values', () => {
            const js = new JavaScript('1 + 1', true, 0, {
                filename: 'test.less'
            });
            expect(js.expression).toBe('1 + 1');
            expect(js.escaped).toBe(true);
            expect(js._index).toBe(0);
            expect(js._fileInfo).toEqual({ filename: 'test.less' });
        });
    });

    describe('eval', () => {
        it('should throw error when javascript is not enabled', () => {
            const js = new JavaScript('1 + 1');
            expect(() => js.eval({ javascriptEnabled: false })).toThrow(
                'Inline JavaScript is not enabled. Is it set in your options?'
            );
        });

        it('should evaluate number expressions and return Dimension', () => {
            const js = new JavaScript('1 + 1');
            const result = js.eval(mockContext);
            expect(result).toBeInstanceOf(Dimension);
            expect(result.value).toBe(2);
        });

        it('should handle NaN values', () => {
            const js = new JavaScript('parseInt("not a number")');
            const result = js.eval(mockContext);
            expect(result).toBeInstanceOf(Anonymous);
            expect(result.value).toBe(NaN);
        });

        it('should evaluate string expressions and return Quoted', () => {
            const js = new JavaScript('"hello" + " world"');
            const result = js.eval(mockContext);
            expect(result).toBeInstanceOf(Quoted);
            expect(result.value).toBe('hello world');
            expect(result.quote).toBe('"');
        });

        it('should evaluate array expressions and return Anonymous with comma-separated values', () => {
            const js = new JavaScript('[1, 2, 3]');
            const result = js.eval(mockContext);
            expect(result).toBeInstanceOf(Anonymous);
            expect(result.value).toBe('1, 2, 3');
        });

        it('should evaluate boolean expressions and return Anonymous', () => {
            const js = new JavaScript('true');
            const result = js.eval(mockContext);
            expect(result).toBeInstanceOf(Anonymous);
            expect(result.value).toBe(true);
        });

        it('should evaluate null and return Anonymous', () => {
            const js = new JavaScript('null');
            const result = js.eval(mockContext);
            expect(result).toBeInstanceOf(Anonymous);
            expect(result.value).toBe(null);
        });

        it('should evaluate undefined and return Anonymous', () => {
            const js = new JavaScript('undefined');
            const result = js.eval(mockContext);
            expect(result).toBeInstanceOf(Anonymous);
            expect(result.value).toBe(undefined);
        });

        it('should handle JavaScript errors gracefully', () => {
            const js = new JavaScript('undefinedVariable');
            expect(() => js.eval(mockContext)).toThrow(
                /JavaScript evaluation error/
            );
        });

        it('should handle syntax errors gracefully', () => {
            const js = new JavaScript('{ invalid syntax }');
            expect(() => js.eval(mockContext)).toThrow(
                /JavaScript evaluation error/
            );
        });

        it('should handle variable interpolation', () => {
            const context = {
                ...mockContext,
                frames: [
                    {
                        variables: () => ({
                            '@color': {
                                value: {
                                    eval: () => ({
                                        toCSS: () => 'red'
                                    }),
                                    toCSS: () => 'red'
                                }
                            }
                        }),
                        variable: (name) => {
                            if (name === '@color') {
                                return {
                                    value: {
                                        eval: () => ({
                                            toCSS: () => 'red'
                                        }),
                                        toCSS: () => 'red'
                                    }
                                };
                            }
                            return null;
                        }
                    }
                ]
            };
            const js = new JavaScript('"@{color}"');
            const result = js.eval(context);
            expect(result).toBeInstanceOf(Quoted);
            expect(result.value).toBe('red');
        });

        it('should handle empty string expressions', () => {
            const js = new JavaScript('');
            expect(() => js.eval(mockContext)).toThrow(
                /JavaScript evaluation error/
            );
        });

        it('should handle special number cases', () => {
            const tests = [
                { expr: 'Infinity', expected: Infinity },
                { expr: '-Infinity', expected: -Infinity },
                { expr: 'Number.MAX_VALUE', expected: Number.MAX_VALUE },
                { expr: 'Number.MIN_VALUE', expected: Number.MIN_VALUE }
            ];

            tests.forEach(({ expr, expected }) => {
                const js = new JavaScript(expr);
                const result = js.eval(mockContext);
                expect(result).toBeInstanceOf(Dimension);
                expect(result.value).toBe(expected);
            });
        });

        it('should handle complex array cases', () => {
            const tests = [
                { expr: '[]', expected: '' },
                { expr: '[1, [2, 3], 4]', expected: '1, 2,3, 4' },
                { expr: '[null, undefined, NaN]', expected: ', , NaN' },
                { expr: '[1, "two", true]', expected: '1, two, true' }
            ];

            tests.forEach(({ expr, expected }) => {
                const js = new JavaScript(expr);
                const result = js.eval(mockContext);
                expect(result).toBeInstanceOf(Anonymous);
                expect(result.value).toBe(expected);
            });
        });

        it('should handle object expressions', () => {
            const js = new JavaScript('({a: 1, b: 2})');
            const result = js.eval(mockContext);
            expect(result).toBeInstanceOf(Anonymous);
            expect(result.value).toEqual({ a: 1, b: 2 });
        });

        it('should handle escaped vs non-escaped strings differently', () => {
            const escaped = new JavaScript('"hello"', true);
            const nonEscaped = new JavaScript('"hello"', false);

            const escapedResult = escaped.eval(mockContext);
            const nonEscapedResult = nonEscaped.eval(mockContext);

            expect(escapedResult.escaped).toBe(true);
            expect(nonEscapedResult.escaped).toBe(false);
        });

        it('should handle multiple context frames with variable precedence', () => {
            const context = {
                ...mockContext,
                frames: [
                    {
                        variables: () => ({
                            '@color': {
                                value: {
                                    eval: () => ({
                                        toCSS: () => 'red',
                                        value: 'red'
                                    }),
                                    toCSS: () => 'red'
                                }
                            }
                        }),
                        variable: (name) => {
                            if (name === '@color') {
                                return {
                                    value: {
                                        eval: () => ({
                                            toCSS: () => 'red',
                                            value: 'red'
                                        }),
                                        toCSS: () => 'red'
                                    }
                                };
                            }
                            return null;
                        }
                    },
                    {
                        variables: () => ({
                            '@color': {
                                value: {
                                    eval: () => ({
                                        toCSS: () => 'blue',
                                        value: 'blue'
                                    }),
                                    toCSS: () => 'blue'
                                }
                            }
                        }),
                        variable: (name) => {
                            if (name === '@color') {
                                return {
                                    value: {
                                        eval: () => ({
                                            toCSS: () => 'blue',
                                            value: 'blue'
                                        }),
                                        toCSS: () => 'blue'
                                    }
                                };
                            }
                            return null;
                        }
                    }
                ]
            };

            const js = new JavaScript('"@{color}"');
            const result = js.eval(context);
            expect(result.value).toBe('red'); // Should use the first frame's value
        });

        it('should provide specific error messages for different error types', () => {
            const syntaxError = new JavaScript('{ invalid syntax }');
            const referenceError = new JavaScript('undefinedVariable');

            try {
                syntaxError.eval(mockContext);
                expect(true).toBe(false); // This will fail if we reach here
            } catch (error) {
                expect(error.message).toContain('Unexpected identifier');
            }

            try {
                referenceError.eval(mockContext);
                expect(true).toBe(false); // This will fail if we reach here
            } catch (error) {
                expect(error.message).toContain('is not defined');
            }
        });
    });
});

File: packages/less/src/less/less_go/javascript_test.go
package less_go

import (
	"strings"
	"testing"
)

// mockFrameTest implements Frame interface for testing
type mockFrameTest struct {
	variables map[string]any
}

func (m *mockFrameTest) Variable(name string) map[string]any {
	if val, ok := m.variables[name]; ok {
		if result, ok := val.(map[string]any); ok {
			return result
		}
	}
	return nil
}

func TestJavaScript(t *testing.T) {
	mockContext := map[string]any{
		"javascriptEnabled": true,
		"frames": []ParserFrame{
			&mockFrameTest{
				variables: map[string]any{},
			},
		},
	}

	t.Run("constructor", func(t *testing.T) {
		// Test file info handling in constructor
		fileInfo := map[string]any{"filename": "test.less"}
		js := NewJavaScript("1 + 1", true, 10, fileInfo)
		if js.expression != "1 + 1" {
			t.Errorf("Expected expression to be '1 + 1', got '%s'", js.expression)
		}
		if !js.escaped {
			t.Error("Expected escaped to be true")
		}
		if js.GetIndex() != 10 {
			t.Errorf("Expected index to be 10, got %d", js.GetIndex())
		}
		if filename, ok := js.FileInfo()["filename"].(string); !ok || filename != "test.less" {
			t.Errorf("Expected filename to be 'test.less', got '%v'", js.FileInfo()["filename"])
		}
	})

	t.Run("eval", func(t *testing.T) {
		t.Run("should throw error when javascript is not enabled", func(t *testing.T) {
			js := NewJavaScript("1 + 1", false, 0, nil)
			_, err := js.Eval(map[string]any{"javascriptEnabled": false})
			if err == nil {
				t.Error("Expected error when javascript is not enabled")
			}
			// Check for the specific error message for disabled JS
			// Use the filename from the node if available, otherwise <unknown>
			expectedMsg := "inline JavaScript is not enabled. Is it set in your options? (filename: <unknown>, index: 0)"
			if err == nil || err.Error() != expectedMsg {
				t.Errorf("Expected specific error message '%s', got '%v'", expectedMsg, err)
			}
		})

		// Helper function to assert the "not supported" error
		assertNotSupportedError := func(t *testing.T, expr string, err error) {
			t.Helper()
			if err == nil {
				t.Errorf("Expected 'JavaScript evaluation is not supported' error for expression '%s', but got nil", expr)
				return
			}
			expectedPrefix := "JavaScript evaluation is not supported in the Go port. Expression: "
			if !strings.HasPrefix(err.Error(), expectedPrefix) {
				t.Errorf("Expected error prefix '%s' for expression '%s', got '%v'", expectedPrefix, expr, err)
			}
		}

		t.Run("should return error for number expressions", func(t *testing.T) {
			expr := "1 + 1"
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should return error for NaN values", func(t *testing.T) {
			expr := "parseInt('not a number')"
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should return error for string expressions", func(t *testing.T) {
			expr := `"hello" + " world"` // Use raw string literal for clarity
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should return error for array expressions", func(t *testing.T) {
			expr := "[1, 2, 3]"
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should return error for boolean expressions", func(t *testing.T) {
			expr := "true"
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should return error for null", func(t *testing.T) {
			expr := "null"
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should return error for undefined", func(t *testing.T) {
			expr := "undefined"
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should handle JavaScript errors gracefully (by returning not supported)", func(t *testing.T) {
			expr := "undefinedVariable"
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should handle syntax errors gracefully (by returning not supported)", func(t *testing.T) {
			expr := "{ invalid syntax }"
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should return error for variable interpolation", func(t *testing.T) {
			context := map[string]any{
				"javascriptEnabled": true,
				"frames": []ParserFrame{
					&mockFrameTest{
						variables: map[string]any{
							"@color": map[string]any{
								"value": NewQuoted("\"", "red", false, 0, nil), // Use Quoted for proper jsify
							},
						},
					},
				},
			}
			expr := `"@{color}"`
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(context)
			// The error here will be slightly different because variable replacement happens first
			expectedPrefix := "JavaScript evaluation is not supported in the Go port. Expression: \"red\""
			if err == nil || !strings.HasPrefix(err.Error(), expectedPrefix) {
				t.Errorf("Expected error prefix '%s', got '%v'", expectedPrefix, err)
			}
		})

		t.Run("should return error for empty string expressions", func(t *testing.T) {
			expr := ""
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should return error for special number cases", func(t *testing.T) {
			tests := []string{
				"Infinity",
				"-Infinity",
				"Number.MAX_VALUE",
				"Number.MIN_VALUE",
			}

			for _, expr := range tests {
				js := NewJavaScript(expr, false, 0, nil)
				_, err := js.Eval(mockContext)
				assertNotSupportedError(t, expr, err)
			}
		})

		t.Run("should return error for complex array cases", func(t *testing.T) {
			tests := []string{
				"[]",
				"[1, [2, 3], 4]",
				"[null, undefined, NaN]",
				`[1, "two", true]`,
			}

			for _, expr := range tests {
				js := NewJavaScript(expr, false, 0, nil)
				_, err := js.Eval(mockContext)
				assertNotSupportedError(t, expr, err)
			}
		})

		t.Run("should return error for object expressions", func(t *testing.T) {
			expr := "({a: 1, b: 2})"
			js := NewJavaScript(expr, false, 0, nil)
			_, err := js.Eval(mockContext)
			assertNotSupportedError(t, expr, err)
		})

		t.Run("should handle escaped vs non-escaped strings differently", func(t *testing.T) {
			escaped := NewJavaScript(`"hello"`, true, 0, nil)
			nonEscaped := NewJavaScript(`"hello"`, false, 0, nil)

			_, err1 := escaped.Eval(mockContext)
			_, err2 := nonEscaped.Eval(mockContext)

			// Both should return not supported errors, but we can verify the escaped flag is preserved
			assertNotSupportedError(t, `"hello"`, err1)
			assertNotSupportedError(t, `"hello"`, err2)
		})

		t.Run("should handle multiple context frames with variable precedence", func(t *testing.T) {
			context := map[string]any{
				"javascriptEnabled": true,
				"frames": []ParserFrame{
					&mockFrameTest{
						variables: map[string]any{
							"@color": map[string]any{
								"value": NewQuoted("\"", "red", false, 0, nil),
							},
						},
					},
					&mockFrameTest{
						variables: map[string]any{
							"@color": map[string]any{
								"value": NewQuoted("\"", "blue", false, 0, nil),
							},
						},
					},
				},
			}

			js := NewJavaScript(`"@{color}"`, false, 0, nil)
			_, err := js.Eval(context)
			
			// The error should contain the value from the first frame (red)
			expectedPrefix := "JavaScript evaluation is not supported in the Go port. Expression: \"red\""
			if err == nil || !strings.HasPrefix(err.Error(), expectedPrefix) {
				t.Errorf("Expected error prefix '%s', got '%v'", expectedPrefix, err)
			}
		})

		// Note: The escaped vs non-escaped test is implicitly covered by checking the error, 
		// as the difference in behavior only matters if evaluation occurs.
	})
} 